### PR TITLE
Overhaul Corporea to pass around CorporeaRequestMatchers instead of Objects

### PR DESCRIPTION
**Pull request by Alwinfy**
Notes:
 - This breaks some API stuff (sorry, getInventoriesWithItemInNetwork), but that's OK because nobody's written a 1.14 addon for Botania anyway
 - This also makes existing Retainers lose their currently-remembered requests (but AFAIK they're broken on HEAD rn anyway)
 - I put the current matchers in static subclasses CorporeaRequestDefaultMatchers for the sake of compactness and `import static` them; if this style isn't OK I can split them up too
 - Yes, I know the serialization mechanic for TileCorporeaRetainer is clunky. But it works, I'm pretty sure. (Couldn't test; see above note about them being broken)

### CODE REVIEW

File: src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java [hunk 1]
```diff
@@ -20,24 +20,20 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import net.minecraftforge.common.MinecraftForge;
+import static vazkii.botania.api.corporea.CorporeaRequestDefaultMatchers.*;
```
Comment (Hubry) [inline src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java]: Static imports go below all other imports

File: src/main/java/vazkii/botania/common/block/tile/corporea/TileCorporeaRetainer.java [hunk 2]
```diff
@@ -10,46 +10,56 @@
...
 
 	private boolean pendingRequest = false;
 	private BlockPos requestPos = BlockPos.ZERO;
-	private Object request;
+	private CorporeaRequestMatcher request;
 	private int requestCount;
 	private int compValue;
+	
+	static {
+		addCorporeaRequestMatcher("string", CorporeaStringMatcher.class,CorporeaStringMatcher::createFromNBT);
```
Comment (Hubry) [inline src/main/java/vazkii/botania/common/block/tile/corporea/TileCorporeaRetainer.java]: formatting

File: src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java [hunk 3]
```diff
@@ -77,57 +73,57 @@
...
 	/**
 	 * Gets the amount of available items in the network of the type passed in, checking NBT or not.
 	 */
-	public static int getCountInNetwork(ItemStack stack, Map<InvWithLocation, Integer> inventories, boolean checkNBT) {
+	public static int getCountInNetwork(CorporeaRequestMatcher matcher, Map<InvWithLocation, Integer> inventories) {
 		int count = 0;
 
-		for(InvWithLocation inv : inventories.keySet())
-			count += inventories.get(inv);
+		for(int value: inventories.values())
```
Comment (Hubry) [inline src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java]: formatting

**Comment by Hubry** [comment]
Also: 
* Would it be worth it to move all the `Map<InvWithLocation, Integer>` to FastUtil `Object2IntMap`s? 
* JEI requesting could probably be moved into a packet or something to request a stack instead of requesting the stack's name. A fairly annoying thing is that you can't request a specific item if two share a name. This could probably go into another PR, I might even do it at some point.

File: src/main/java/vazkii/botania/api/corporea/CorporeaRequestMatcher.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	 */
+	public boolean isStackValid(ItemStack stack) {
+		return false;
+	}
+
+	/**
+	 * Serialize to NBT data, for the Corporea Retainer's benefit.
+	 * No-op; should be overridden by subclasses.
+	 */
+	public void writeToNBT(CompoundNBT tag) {}
```
Comment (williewillus) [inline src/main/java/vazkii/botania/api/corporea/CorporeaRequestMatcher.java]: if these methods "should all be overridden by subclasses", just make them abstract to force implementation. no point giving a noop impl that could potentially be missed. Also make this an interface maybe?

Comment (Alwinfy) [reply src/main/java/vazkii/botania/api/corporea/CorporeaRequestMatcher.java]: Did that, thanks.

**Pull request merged**